include/deal.II/fe/fe_series.h

```diff
@@ -0,0 +1,428 @@
+
+
+
+  /**
+   * Linear regression least-square fit of $y = k \, x + b$.
+   * The size of the input vectors should be equal and more than 1.
+   * The returned pair will contain $k$ (first) and $b$ (second).
+   */
+  std::pair<double,double> linear_regression(const std::vector<double> &x,
+                                             const std::vector<double> &y);
+
+}
+
+/*@}*/
+
+#ifndef DOXYGEN
+
+// -------------------  inline and template functions ----------------
+
+namespace
+{
+  template <int dim,typename T>
+  void fill_map_index(const Table<dim,T> &coefficients,
+                      const TableIndices<dim> &ind,
+                      const std_cxx11::function<std::pair<bool,unsigned int>(const TableIndices<dim> &)> &predicate,
+                      std::map<unsigned int, std::vector<T> > &pred_to_values)
+  {
+    const std::pair<bool,unsigned int> pred_pair = predicate(ind);
+    // don't add a value if predicate is false
+    if (pred_pair.first == false)
+      return;
+
+    const unsigned int &pred_value = pred_pair.second;
+    const T &coeff_value = coefficients(ind);
+
+    if (pred_to_values.find (pred_value) == pred_to_values.end())
+      {
+        pred_to_values[pred_value] = std::vector<T>(1,coeff_value);
+      }
+    else
+      {
+        pred_to_values[pred_value].push_back(coeff_value);
+      }
+  }
+
+  template <typename T>
+  void fill_map(const Table<1,T> &coefficients,
+                const std_cxx11::function<std::pair<bool,unsigned int>(const TableIndices<1> &)> &predicate,
+                std::map<unsigned int, std::vector<T> > &pred_to_values)
+  {
+    for (unsigned int i = 0; i < coefficients.size(0); i++)
+      {
+        const TableIndices<1> ind(i);
+        fill_map_index(coefficients,ind,predicate,pred_to_values);
+      }
+
+  }
+
+  template <typename T>
+  void fill_map(const Table<2,T> &coefficients,
+                const std_cxx11::function<std::pair<bool,unsigned int>(const TableIndices<2> &)> &predicate,
+                std::map<unsigned int, std::vector<T> > &pred_to_values)
+  {
+    for (unsigned int i = 0; i < coefficients.size(0); i++)
+      for (unsigned int j = 0; j < coefficients.size(1); j++)
+        {
+          const TableIndices<2> ind(i,j);
+          fill_map_index(coefficients,ind,predicate,pred_to_values);
+        }
+
+  }
+
+  template <typename T>
+  void fill_map(const Table<3,T> &coefficients,
+                const std_cxx11::function<std::pair<bool,unsigned int>(const TableIndices<3> &)> &predicate,
+                std::map<unsigned int, std::vector<T> > &pred_to_values)
+  {
+    for (unsigned int i = 0; i < coefficients.size(0); i++)
+      for (unsigned int j = 0; j < coefficients.size(1); j++)
+        for (unsigned int k = 0; k < coefficients.size(2); k++)
+          {
+            const TableIndices<3> ind(i,j,k);
+            fill_map_index(coefficients,ind,predicate,pred_to_values);
+          }
```

Comment:
i think the problem here is the nested loops which depend on the dimension. It would be good to have a `.begin()` `.end()` iterator for `Table` with `it->first` being `TableIndices` and `it->second` being coefficient values, but this is outside of the scope of this PR.